internal/graph/check_fast_path.go

```diff
@@ -0,0 +1,674 @@
+				batch = make([]*openfgav1.TupleKey, 0)
+			}
+			var err error
+			iterProducers, err = pollIteratorProducers(ctx, iterProducers)
+			if err != nil {
+				concurrency.TrySendThroughChannel(ctx, &iteratorMsg{err: err}, iterChan)
+				return
+			}
+		}
+	}
+}
+
+func (c *LocalChecker) fastPathOperationSetup(ctx context.Context, req *ResolveCheckRequest, op setOperatorType, children ...*openfgav1.Userset) (chan *iteratorMsg, error) {
+	iterProducers := make([]*iteratorProducer, 0, len(children))
+	for idx, child := range children {
+		producerChan, err := c.fastPathRewrite(ctx, req, child)
+		if err != nil {
+			return nil, err
+		}
+		iterProducers = append(iterProducers, &iteratorProducer{idx: idx, producer: producerChan})
+	}
+	var resolver fastPathSetHandler
+	switch op {
+	case unionSetOperator:
+		resolver = fastPathUnion
+	case intersectionSetOperator:
+		resolver = fastPathIntersection
+	case exclusionSetOperator:
+		resolver = fastPathDifference
+	default:
+		return nil, ErrUnknownSetOperator
+	}
+	resultChan := make(chan *iteratorMsg, len(children))
+	go resolver(ctx, iterProducers, resultChan)
+	return resultChan, nil
+}
+
+func (c *LocalChecker) fastPathRewrite(
+	ctx context.Context,
+	req *ResolveCheckRequest,
+	rewrite *openfgav1.Userset,
+) (chan *iteratorMsg, error) {
+	switch rw := rewrite.GetUserset().(type) {
+	case *openfgav1.Userset_This:
+		return c.fastPathDirect(ctx, req)
+	case *openfgav1.Userset_ComputedUserset:
+		return c.fastPathComputed(ctx, req, rewrite)
+	case *openfgav1.Userset_Union:
+		return c.fastPathOperationSetup(ctx, req, unionSetOperator, rw.Union.GetChild()...)
+	case *openfgav1.Userset_Intersection:
+		return c.fastPathOperationSetup(ctx, req, intersectionSetOperator, rw.Intersection.GetChild()...)
+	case *openfgav1.Userset_Difference:
+		return c.fastPathOperationSetup(ctx, req, exclusionSetOperator, rw.Difference.GetBase(), rw.Difference.GetSubtract())
+	default:
+		return nil, ErrUnknownSetOperator
+	}
+}
+
+// resolveFastPath attempts to find the intersection across 2 producers (channels) of ObjectIDs.
+// In the case of a TTU:
+// Right channel is the result set of the Read of ObjectID/Relation that yields the User's ObjectID.
+// Left channel is the result set of ReadStartingWithUser of User/Relation that yields Object's ObjectID.
+// From the perspective of the model, the left hand side of a TTU is the computed relationship being expanded.
+func (c *LocalChecker) resolveFastPath(ctx context.Context, leftChans []chan *iteratorMsg, iter TupleMapper) (*ResolveCheckResponse, error) {
+	rightOpen := true
+	leftOpen := true
+
+	cancellableCtx, cancel := context.WithCancel(ctx)
+	leftChan := fanInIteratorChannels(cancellableCtx, leftChans)
+	rightChan := streamedLookupUsersetFromIterator(cancellableCtx, iter)
+
+	defer func() {
+		cancel()
+		iter.Stop()
+		if !leftOpen {
+			return
+		}
+		go func() {
+			for msg := range leftChan {
+				if msg.iter != nil {
+					msg.iter.Stop()
+				}
+			}
+		}()
```

Comment:
1. Nitpick: you don't need the `if !leftOpen return` because the ranging over a channel is a no-op if the channel is closed. https://go.dev/play/p/A9w8ThwGg54
2. Why drain the channel over a goroutine? This doesn't look like an expensive operation, what am i missing?

Reply:
1. Not needed if I weren't launching a goroutine. But I want to avoid launching it at all costs.
2. May not look as an expensive operation if you think everything is on memory. But the reality is that you don't know if the producer is one that is yielding over the network and has yet to properly cancel. Also think of the case where you had 100s iterators queued, you have to close them all. In light of not knowing how many there are and prevent increasing the latency to reply to the customer, we simply launch the goroutine.